src/drivers/imx/micfil.c

```diff
@@ -0,0 +1,336 @@
+
+/* dd400475-35d7-4045-ab03-0c34957d7a08 */
+DECLARE_SOF_UUID("micfil-dai", micfil_uuid, 0xdd400475, 0x35d7, 0x4045,
+		 0xab, 0x03, 0x0c, 0x34, 0x95, 0x7d, 0x7a, 0x08);
+
+DECLARE_TR_CTX(micfil_tr, SOF_UUID(micfil_uuid), LOG_LEVEL_INFO);
+
+#define MICFIL_OSR_DEFAULT		16
+/* set default gain to 2 */
+#define MICFIL_DEFAULT_ADJ_RANGE	0x22222222
+#define MICFIL_CLK_ROOT			24576000
+
+enum micfil_quality {
+	QUALITY_HIGH,
+	QUALITY_MEDIUM,
+	QUALITY_LOW,
+	QUALITY_VLOW0,
+	QUALITY_VLOW1,
+	QUALITY_VLOW2,
+};
+
+static int micfil_reset(struct dai *dai)
+{
+	dai_update_bits(dai, REG_MICFIL_CTRL1, MICFIL_CTRL1_MDIS, 0);
+	dai_update_bits(dai, REG_MICFIL_CTRL1, MICFIL_CTRL1_SRES, MICFIL_CTRL1_SRES);
+	dai_update_bits(dai, REG_MICFIL_STAT, 0xff, 0xff);
+
+	return 0;
+}
+
+static int micfil_get_hw_params(struct dai *dai,
+				struct sof_ipc_stream_params *params, int dir)
+{
+	struct micfil_pdata *micfil = dai_get_drvdata(dai);
+
+	dai_err(dai, "micfil_get_hw_params()");
+
+	params->rate = micfil->params.pdm_rate;
+	params->channels = micfil->params.pdm_ch;
+	params->buffer_fmt = SOF_IPC_BUFFER_INTERLEAVED;
+	params->frame_fmt = SOF_IPC_FRAME_S32_LE;
+
+	return 0;
+}
+
+static int micfil_set_quality(struct dai *dai)
+{
+	struct micfil_pdata *micfil = dai_get_drvdata(dai);
+	int qsel;
+
+	switch (micfil->quality) {
+	case QUALITY_HIGH:
+		qsel = MICFIL_QSEL_HIGH_QUALITY;
+		break;
+	case QUALITY_MEDIUM:
+		qsel = MICFIL_QSEL_MEDIUM_QUALITY;
+		break;
+	case QUALITY_LOW:
+		qsel = MICFIL_QSEL_LOW_QUALITY;
+		break;
+	case QUALITY_VLOW0:
+		qsel = MICFIL_QSEL_VLOW0_QUALITY;
+		break;
+	case QUALITY_VLOW1:
+		qsel = MICFIL_QSEL_VLOW1_QUALITY;
+		break;
+	case QUALITY_VLOW2:
+		qsel = MICFIL_QSEL_VLOW2_QUALITY;
+		break;
+	default:
+		dai_err(dai, "MICFIL: invalid quality mode %d", micfil->quality);
+		return -EINVAL;
+	}
+
+	dai_update_bits(dai, REG_MICFIL_CTRL2, MICFIL_CTRL2_QSEL,
+			MICFIL_CTRL2_QSEL_BITS(qsel));
+
+	return 0;
+}
+
+/* get_pdm_clk - computes the product between k-factor and PDM_CLK rate
+ * @param dai, SOF DAI struct
+ * @param rate, output sampling rate
+ *
+ * PDM_CLK depends on Quality Mode, output sampling rate and quality
+ * mode
+ */
+static int micfil_get_pdm_clk(struct dai *dai, int rate)
+{
+	unsigned int osr;
+	unsigned int qsel;
+	unsigned int ctrl2_reg;
+	unsigned int pdm_clk = 0;
+
+	ctrl2_reg = dai_read(dai, REG_MICFIL_CTRL2);
+	osr = 16 - ((ctrl2_reg & MICFIL_CTRL2_CICOSR) >> MICFIL_CTRL2_CICOSR_SHIFT);
+	qsel = (ctrl2_reg & MICFIL_CTRL2_QSEL) >> MICFIL_CTRL2_QSEL_SHIFT;
+
+	switch (qsel) {
+	case MICFIL_QSEL_HIGH_QUALITY:
+		pdm_clk = rate * 8 * osr / 2; /* kfactor = 0.5 */
+		break;
+	case MICFIL_QSEL_MEDIUM_QUALITY:
+	case MICFIL_QSEL_VLOW0_QUALITY:
+		pdm_clk = rate * 4 * osr; /* kfactor = 1 */
+		break;
+	case MICFIL_QSEL_LOW_QUALITY:
+	case MICFIL_QSEL_VLOW1_QUALITY:
+		pdm_clk = rate * 2 * osr * 2; /* kfactor = 2 */
+		break;
+	case MICFIL_QSEL_VLOW2_QUALITY:
+		pdm_clk = rate * osr; /* kfactor = 4 */
```

Comment:
can any of the above overflow to become negative?

Reply:
it shouldn't. We only support rates under 48000 and osr can be up to 256. I guess this function should return an unsigned value.